### PR TITLE
interfaces: scala typechecker implementation

### PR DESCRIPTION
**Pull request by robin-da**
This is the scala side of the lf typechecker for interfaces #10810 .

There's also a cosmetic change in the haskell typechecker I corrected.

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala [hunk 1]
```diff
@@ -431,6 +454,28 @@ private[validation] object Typing {
...
+    }
+
+    def checkIfaceChoice(choice: InterfaceChoice): Unit = {
+      checkType(choice.argType, KStar)
+      checkType(choice.returnType, KStar)
+    }
+
+    def checkIfaceImplementation(tplTcon: TypeConName, ifaceTcon: TypeConName): Unit = {
+      if (
+        !(tplTcon.packageId == ifaceTcon.packageId && tplTcon.qualifiedName.module == ifaceTcon.qualifiedName.module)
```
Comment (cocreature) [inline daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala:467]: this also seems off

**Review by cocreature — APPROVED** [review]
Mostly looks good except for the package/module checks. We should probably do the LF parser soon so we can test this nicely.

**Review by remyhaemmerle-da — APPROVED** [review]
Looks good to me, except we need a special the "context" of `lookupTemplateOrInterface`.

Could you please add TODOs in test for lookups, collision, and types?

File: daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/Interface.scala [hunk 2]
```diff
@@ -179,6 +179,27 @@ private[lf] class Interface(signatures: PartialFunction[PackageId, PackageSignat
...
+
+  def lookupInterface(name: TypeConName): Either[LookupError, DefInterface] =
+    lookupInterface(name, Reference.Interface(name))
+
+  def lookupTemplateOrInterface(
+      name: TypeConName
+  ): Either[LookupError, Either[TemplateSignature, DefInterface]] = {
+    val x = lookupTemplate(name).map(Left[TemplateSignature, DefInterface](_))
+    val y = lookupInterface(name).map(Right[TemplateSignature, DefInterface](_))
+    x.fold(_ => y, _ => x)
```
[the remark this answers is not in the document]
Comment (robin-da) [reply daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/Interface.scala]: this has been purged now with the rebase to separate primitives.

[the remark this answers is not in the document]
Comment (robin-da) [reply daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/Interface.scala]: (but I really felt like a scala pro there for a second)

[the remark this answers is not in the document]
Comment (cocreature) [reply daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/Interface.scala]: you’ll always be a scala pro in our hearts :heart:

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/ValidationError.scala [hunk 3]
```diff
@@ -409,3 +416,44 @@ final case class EModuleVersionDependencies(
 
   override def context: Context = NoContext
 }
+final case class EForeignInterfaceImplementation(
+    context: Context,
+    iface: TypeConName,
+) extends ValidationError {
+
+  override protected def prettyInternal: String =
+    s"The definition and implementation for the interface $iface need to be in the same module."
```
[the remark this answers is not in the document]
Comment (robin-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/ValidationError.scala:425]: Fixed.

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/ValidationError.scala [hunk 4]
```diff
@@ -409,3 +416,44 @@ final case class EModuleVersionDependencies(
...
+
+final case class EBadInterfaceChoiceImplArgType(
+    context: Context,
+    choice: ChoiceName,
+    ifaceArgType: Type,
+    tplArgType: Type,
+) extends ValidationError {
+
+  override protected def prettyInternal: String =
+    s"Choice implementation and interface definition for $choice differ in argument type.\nExpected: $ifaceArgType\n But got: $tplArgType"
```
[the remark this answers is not in the document]
Comment (robin-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/ValidationError.scala:447]: fixed.

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/ValidationError.scala [hunk 5]
```diff
@@ -409,3 +416,44 @@ final case class EModuleVersionDependencies(
...
+
+final case class EBadInterfaceChoiceImplRetType(
+    context: Context,
+    choice: ChoiceName,
+    ifaceRetType: Type,
+    tplRetType: Type,
+) extends ValidationError {
+
+  override protected def prettyInternal: String =
+    s"Choice implementation and interface defintion for $choice differ in return type.\nExpected: $ifaceRetType\nBut got: $tplRetType"
```
[the remark this answers is not in the document]
Comment (robin-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/ValidationError.scala:458]: fixed.

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/ValidationError.scala [hunk 6]
```diff
@@ -409,3 +416,44 @@ final case class EModuleVersionDependencies(
...
+
+final case class EBadInterfaceChoiceImplConsuming(
+    context: Context,
+    choice: ChoiceName,
+    ifaceConsuming: Boolean,
+    tplConsuming: Boolean,
+) extends ValidationError {
+
+  override protected def prettyInternal: String =
+    s"Choice implementation and interface definition for $choice differ in consuming/non-consuming behaviour.\nExpected: $ifaceConsuming\nBut got: $tplConsuming\n"
```
[the remark this answers is not in the document]
Comment (robin-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/ValidationError.scala]: done

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs [hunk 7]
```diff
@@ -836,19 +836,18 @@ checkTemplate m t@(Template _loc tpl param precond signatories observers text ch
...
-  => ModuleName
-  -> Qualified TypeConName
+  => Qualified TypeConName
   -> Qualified TypeConName
   -> m ()
-checkIfaceImplementation modName tplTcon ifTcon
-  | Qualified PRSelf m _tconName <- ifTcon
-  , m == modName = do
+checkIfaceImplementation tplTcon ifTcon
+  | qualPackage tplTcon == qualPackage ifTcon
```
[the remark this answers is not in the document]
Comment (robin-da) [reply compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs:855]: I think I bonked this, this should be a check that the template and the interface implementation live in the same module, but that's a given anyways the way we implement it now. I'll remove the check and error.

File: daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/Interface.scala [hunk 8]
```diff
@@ -179,6 +179,27 @@ private[lf] class Interface(signatures: PartialFunction[PackageId, PackageSignat
...
+    lookupModule(name.packageId, name.qualifiedName.module, context).flatMap(
+      _.interfaces
+        .get(name.qualifiedName.name)
+        .toRight(LookupError(Reference.Interface(name), context))
+    )
+
+  def lookupInterface(name: TypeConName): Either[LookupError, DefInterface] =
+    lookupInterface(name, Reference.Interface(name))
+
+  def lookupTemplateOrInterface(
```
[the remark this answers is not in the document]
Comment (robin-da) [reply daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/Interface.scala]: This got removed with the seperate primitives now.

File: daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/Interface.scala [hunk 9]
```diff
@@ -179,6 +179,27 @@ private[lf] class Interface(signatures: PartialFunction[PackageId, PackageSignat
...
+      name: TypeConName
+  ): Either[LookupError, Either[TemplateSignature, DefInterface]] = {
+    val x = lookupTemplate(name).map(Left[TemplateSignature, DefInterface](_))
+    val y = lookupInterface(name).map(Right[TemplateSignature, DefInterface](_))
```
[the remark this answers is not in the document]
Comment (robin-da) [reply daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/Interface.scala]: also got removed with the separate primitives for exerciseInterface/fetchInterface.

**Comment by robin-da** [comment]
@remyhaemmerle-da I added todos for typing/collision specs.

**Pull request merged**